File: algorithms-java/src/main/java/me/freedom4live/algorithms/factors/SimpleFactorCalculator.java
package me.freedom4live.algorithms.factors;

import java.util.ArrayList;
import java.util.List;

class SimpleFactorCalculator {

    static List<Integer> calculate(Integer number) {
        List<Integer> result = new ArrayList<>();

        number = checkOnTwo(number, result);
        checkOtherFactors(number, result);

        return result;
    }

    private static void checkOtherFactors(Integer number, List<Integer> result) {
        int maxFactor = (int) Math.sqrt(number);
        int currentFactor = 3;

        while (currentFactor <= maxFactor) {
            while (number % currentFactor == 0) {
                result.add(currentFactor);

                number = number / currentFactor;
                maxFactor = (int) Math.sqrt(number);
            }

            currentFactor += 2;
        }

        if (number > 1) result.add(number);
    }

    private static Integer checkOnTwo(Integer number, List<Integer> result) {
        while (number % 2 == 0) {
            result.add(2);
            number = number / 2;
        }
        return number;
    }

}

File: algorithms-kotlin/src/main/kotlin/me/freedom4live/algorithms/factors/SimpleFactorCalculator.kt
package me.freedom4live.algorithms.factors

object SimpleFactorCalculator {

    fun calculate(number: Int): List<Int> = Result(number).run {
        findFactorOfTwoAndUpdateResult(this)
        findFactorsOfOtherNumbersAndUpdateResult(this)
        findAdditionalFactorAndUpdateResult(this)
        this.getFactorsAsList()
    }

    private fun findFactorOfTwoAndUpdateResult(result: Result) {
        while (result.reminder % 2 == 0) {
            result.addFactor(2)
            result.reminder /= 2
        }
    }

    private fun findFactorsOfOtherNumbersAndUpdateResult(result: Result) {
        var currentFactor = 3
        var maxFactor: Int = computeCurrentMaxFactor(result)

        while (maxFactor >= currentFactor) {
            checkNumberByFactor(result, currentFactor)
            maxFactor = computeCurrentMaxFactor(result)

            currentFactor += 2
        }
    }

    private fun findAdditionalFactorAndUpdateResult(result: Result) {
        if (result.reminder > 1) result.addFactor(result.reminder)
    }

    private fun checkNumberByFactor(result: Result, currentFactor: Int) {
        while (result.reminder % currentFactor == 0) {
            result.addFactor(currentFactor)
            result.reminder /= currentFactor
        }
    }

    private fun computeCurrentMaxFactor(result: Result) =
        Math.sqrt(result.reminder.toDouble()).toInt()

    private data class Result(var reminder: Int) {
        private val factors: MutableList<Int> = mutableListOf()

        fun addFactor(factor: Int) {
            factors.add(factor)
        }

        fun getFactorsAsList() = factors
    }

}
